src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -506,63 +509,60 @@ public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompl
         public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption,
             CancellationToken cancellationToken)
         {
+            // Called outside of async state machine to propagate certain exception even without awaiting the returned task.
+            CheckRequestBeforeSend(request);
+
             return SendAsyncCore(request, completionOption, async: true, cancellationToken).AsTask();
         }
 
-        private ValueTask<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption,
-            bool async, CancellationToken cancellationToken)
+        private void CheckRequestBeforeSend(HttpRequestMessage request)
         {
             if (request == null)
             {
                 throw new ArgumentNullException(nameof(request));
             }
             CheckDisposed();
-            CheckRequestMessage(request);
 
             SetOperationStarted();
-            PrepareRequestMessage(request);
+
             // PrepareRequestMessage will resolve the request address against the base address.
+            PrepareRequestMessage(request);
+        }
 
+        private async ValueTask<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption,
```

Comment:
This is now introducing an extra async state machine into every request, regardless of whether tracing is enabled.

Reply:
It shouldn't, unless I've overlooked something.
I merged it with `FinishSendAsync` which already was async, the number of async state machines should be the same.

Reply:
Ah, you made this async but then removed an async helper it was calling, so the actual number of async methods remains the same?

Reply:
Ok

Reply:
Yep, exactly.